[15.0] connector_importer: misc fix/imp

simahawk opened the pull request:
See atomic commits for details.
Basically moving out of product importer work branch here https://github.com/camptocamp/connector-interfaces/tree/15-mig-connector_importer_product

Need to get back to this for review and tests.

simahawk committed:
connector_importer: add dynamic mapper

simahawk committed:
connector_importer: improve xid handling

simahawk committed:
connector_importer: fix logger.warn deprecated

simahawk committed:
connector_importer: mapper_utils store from key

This way we can inspect the converter function
to retrieve the original key.

simahawk committed:
connector_importer: support multi valued xids

simahawk committed:
c_importer: backend_to_rel support Odoo studio

simahawk committed:
connector_importer: add import type view

simahawk committed:
connector_importer: fix importer.record apply_on

The importer for a record should be model agnostic
otherwise we cannot use it for any other odoo model.

This is very important to support custom mappers from settings
w/out specific importers.

simahawk committed:
connector_importer: allow custom mapper from conf

You can now declare the mapper to be used via import.type options.

simahawk committed:
connector_importer: improve dynamic mapper

You can now pass via import type options:

* source_key_prefix -> to filter all keys via specific prefix

  This is really handy when in your source (normally a CSV)
  you have a set of fields to be used for creating/updating other records

* source_key_whitelist -> consider only these specific keys from the source

The prefix will be cleaned up automatically and what remains after it
must be a real Odoo field on the current model.

simahawk committed:
connector_importer: support 'override_existing' from conf

You can now pass override_existing flag in the importer options.

simahawk committed:
connector_importer: allow mapper settings from conf

You can now pass:

* required_keys
* translatable_keys
* default_keys

in the 'mapper' section of import type options.

Eg:

- model: foo
  importer:
    name: bla
  options:
    mapper:
      name: my.mapper
      required_keys:
          source_key: dest_key
          __source_key2: dest_key2
      default_keys:
        field1: true

simahawk committed:
connector_importer: unify importer lookup by type conf

simahawk committed:
connector_importer: fix recordset name compute

Addig the name of the backend is useless and can be broken
till you have a backend.

simahawk committed:
connector_importer: fix recordset docs compute

simahawk committed:
connector_importer: add source lines slice option

You can decide to work on a specific range of CSV lines.

simahawk committed:
connector_importer: add source csv tree + menu

simahawk committed:
connector_importer: add import type desc

simahawk committed:
c_importer: dynamic mapper add source_key_empty_skip

New config key to allow skipping fields when they are empty.

Use cases:

  * field w/ unique constraint but not populated (eg: product barcode)
  * field not to override when empty

dhx reviewed (approved):
Did some basic tests - LGTM

simahawk committed:
connector_importer: dynamicmapper tests + fixes

simahawk commented:
pre-commit fails like

```
   File "/home/runner/.cache/pre-commit/repo3jub9f__/py_env-python3/lib/python3.11/site-packages/wrapt/__init__.py", line 10, in <module>
    from .decorators import (adapter_factory, AdapterFactory, decorator,
  File "/home/runner/.cache/pre-commit/repo3jub9f__/py_env-python3/lib/python3.11/site-packages/wrapt/decorators.py", line 34, in <module>
    from inspect import ismethod, isclass, formatargspec
ImportError: cannot import name 'formatargspec' from 'inspect' (/opt/hostedtoolcache/Python/3.11.0/x64/lib/python3.11/inspect.py)

pylint with mandatory checks.............................................Failed
- hook id: pylint
- exit code: 1
```
I'm trying a template sync here #97

simahawk committed:
c_importer: transl use regional lang False by default

simahawk committed:
connector_importer: validate translatable keys

Translatable keys might refer non translatable fields
or not existing fields: purge them if needed
and log proper errors.

sebalix commented on connector_importer/components/mapper.py:
```diff
@@ -61,7 +66,23 @@ def translatable_keys(self, create=False):
+        translatable = self._validate_translate_keys(set(translatable))
+        return tuple(translatable)
+
+    def _validate_translate_keys(self, translatable):
```
:+1:

sebalix commented on connector_importer/models/recordset.py:
```diff
@@ -272,24 +271,21 @@ def generate_report(self):
         for item in self:
             item.docs_html = False
-            if isinstance(item.id, models.NewId):
+            if isinstance(item.id, models.NewId) or not item.backend_id:
+                # Surprise surprise: when editing a new recordset
+                # if you hit `configure source` btn
+                # the record will be saved but the backend can be null :S
```
Not sure to understand, if you create a new recordset from the UI, as there is no `backend_id` on the form it is expected to not have it defined here no?

sebalix reviewed (approved):
Approving anyway, LG

simahawk commented:
/ocabot merge minor
